### DISCUSSION TITLE
Multi-row tab bar

### DISCUSSION THREAD

**Discussion by yurikhan** [minbrowser/min#1744] [Ideas]
Hi! I’m a long-time Firefox user and getting a bit fed up with Mozilla’s policies, so looking around for a browser that could accommodate my habits that were formed by pre-WebExtension Firefox. (Namely, there was an extension called Tab Mix Plus, that had a number of radical ideas regarding tabs.)

One such habit is tab hoarding. The window I’m currently typing in has 43 tabs. I have 24 more windows with 20, 19, 13, 12, 8, 7, 7, 7, 6, 6, 6, 5, 4, 4, 3, 3, 3, 2, 1, 1, 1, 1, 1, and 1 tabs. (Oh sh*t, I didn’t realize I had so many. Should probably close some of those.)

Anyway, on to the point of the post. When you have that many tabs in a single window, depending on the browser, you get one of the following:

* Ridiculously small tabs where only part of the favicon or just a few letters of the title fits.
* Horizontal scrolling in the tab bar. This way, you see the current tab and a few tabs around, but you don’t see all the rest. Because you don’t see them, you’re likely to open near duplicates, compounding the problem.
* Multi-row tab bar. I can tell there is a non-negligible share of people who are used to this, but no popular browser supports it. In Firefox, we have to apply a userChrome.css hack that Mozilla keeps breaking.

![image](https://user-images.githubusercontent.com/238691/133932768-097ee2e1-23b4-4d5f-b06f-a827be7d89e0.png)

Would Min, in principle, like to cater to users like me?

(I took a preliminary look at the code, just to see if that’s doable. Seems possible, although it raises some questions as to how a multi-row tab bar interacts with window control buttons. Also, the tab bar would then have dynamic height and that would need some logic to position the webviews.)

**Comment by PalmerAL** [comment]
In general, we do want to support having large numbers of tabs open, although I'm convinced a multi-row tab bar is a good solution to that. The main idea behind tab management in Min is that you split your tabs into smaller/more manageable groups from the tasks menu (in the top right), and then you can switch between them and keep your tabs better organized than if they were all in a single tab bar. It's kinda similar to the old tab groups feature in Firefox (although without the visual organization aspect), if you've ever used that.

Also, have you ever tried tree-style tabs or another vertical tabs extension? What do you prefer about a multi-row tab bar compared to that?

**Reply by yurikhan** [reply]
Tab grouping does look like a feature I’d like to use. I currently group tabs by arranging them in separate Firefox windows, but I cannot assign names to windows, so when I need to switch to a task I have to identify its window by contained tabs, and that’s a cognitive load.

That said, a task can still have more tabs than can fit comfortably in a single row. I use a 24″ monitor and at this size it makes sense to tile two windows, e.g. browser and code editor. This leaves me with browser windows 960px wide. A single tab needs at least 100px so that a little bit of the tab title is visible so I can identify them. Therefore, at most 9 tabs fit in a single row (fewer if we account for all the buttons).

The above also explains why I cannot use a vertical tab bar — it would take away from the window width and the remaining space would not be sufficient for most of the sites out there. Even with my current setup, many sites assume at least 1024px are available and I have to either inject custom CSS or zoom out to make them fit.

This is why I use a multi-row tab bar: it lets me see all the tabs in a task at once, and it only takes a few lines at the top of my half-screen, where I can afford it.

The task overlay in Min is a vertical solution that does not take away from page width. With current styling, I reckon it can display about 20 tabs without scrolling; if I settle on the task overlay, I’d probably try to make that more compact. It also takes multiple clicks to switch tabs. Could probably use some keyboard accessibility improvements — in particular, some kind of search or filtering.

**Reply by PalmerAL** [reply]
That makes sense; thanks.

I agree on the keyboard shortcuts; we should have some kind of search at the top of the overlay.
